Adding SystemController interface design doc (#1224)
kammce/SJSU-Dev2: documentation/design_docs/L1/system_controller.md

be required to map IDs to the appropriate registers.

# Detailed Design

## Interface

```C++
class SystemController
{
 private:
  static inline SystemController * platform_system_controller = nullptr;

 public:
  static void SetPlatformController(SystemController * system_controller);
  static SystemController & GetPlatformController();

  struct ResourceID
  {
    template <size_t id>
    static constexpr ResourceID Define();

    uint32_t device_id = -1;

    bool operator==(const ResourceID & compare) const;
  };

  virtual Returns<void> Initialize() = 0;
  virtual void * GetClockConfiguration() = 0;
  virtual Returns<units::frequency::hertz_t> GetClockRate(
      ResourceID resource) const = 0;
  virtual Returns<bool> IsPeripheralPoweredUp(ResourceID peripheral) const = 0;
  virtual Returns<void> PowerUpPeripheral(ResourceID peripheral) const = 0;
  virtual Returns<void> PowerDownPeripheral(ResourceID peripheral) const = 0;

  template <class ClockConfiguration>
  ClockConfiguration & GetClockConfiguration();
};
```

## Configuring the Default System Controller for the Platform

### static void SetPlatformController(SystemController * system_controller)

> Sets the specified system controller for the platform.

Reviewer: ```suggestion
Sets the specified system controller for the platform. Each platform is responsible for calling this function correctly from within the `InitializePlatform()` function. Without doing so, access to `GetPlatformController()` will return a nullptr.
```

### static SystemController & GetPlatformController()

> Returns the current system controller for the platform.

Reviewer: ```suggestion
Returns the current system controller for the platform. Note that this is used by peripheral drivers to power themselves on and determine their running clock rate. This is also used for application developers to set up "...". This should not be used by L2 HAL or L3 Application as knowledge about the particular platform is not and should not be knowable at those levels.
```
Or something along those lines.

## Clock Configuration

### void * GetClockConfiguration()

> Returns a generic pointer to the system controller's `ClockConfiguration`.

Reviewer: ```suggestion
Returns a pointer to the memory location of system controller's `ClockConfiguration`. In generally this should be called directly. Prefer to call `ClockConfiguration & GetClockConfiguration()` instead
```

    - [Returns&lt;void> PowerDownPeripheral(PeripheralID peripheral) const](#returnsvoid-powerdownperipheralperipheralid-peripheral-const)
- [Future Advancements](#future-advancements)
- [Testing Plan](#testing-plan)
  - [Unit Testing Scheme](#unit-testing-scheme)

# Location
`L1 Peripheral`

# Type
`Interface`

# Background
A SystemController manages a platform's:

  - Clocks and their speeds
  - PLLs feeding into the system's clocks
  - Manages power systems of peripherals and system blocks

The SystemController also gives information about the system such as what
speed a clock is running at so that their peripheral drivers can calculate
things such as the clock rates for serial communicate or the clock cycles before
an event or interrupt is planned to occur.

# Overview
Every SystemController should take a custom `ClockConfiguration` data
structure reference as an input in their constructor, unless a system does
not incorporate any form of clock tree.

> See [Clock Control Guide](https://sjsu-dev2.readthedocs.io/en/latest/guides/clock_systems)
> for more information regarding the use of `ClockConfiguration`.

> The SystemController uses the `PeripheralID` helper class to define various

Author: Would it be better to name **PeripheralID** as **ResourceID**?

Reviewer: Yeah, that sounds like a good generic name for that. Good idea!
